File: src/ipc/journal.ts
/**
 * Typed wrapper over a systemd unit's own recent journal lines.
 *
 * Read-only, matching `ssh/journal.rs`'s own scope: nothing here rotates,
 * vacuums or otherwise touches the journal, only reads its tail.
 */

import { invoke } from '@tauri-apps/api/core';

import type { SessionHandle } from './sessions';

/**
 * The last lines `journalctl` reports for `unit`, in the order the host
 * printed them.
 *
 * A unit whose journal cannot be read (a permission error, a unit name that
 * does not look real, or a host with no `journalctl` at all) yields an
 * empty list rather than a rejection: see `ssh/journal.rs`'s own doc
 * comment for why the three are indistinguishable from here on purpose.
 */
export async function sessionUnitJournal(
  handle: SessionHandle,
  unit: string,
): Promise<readonly string[]> {
  return invoke<string[]>('session_unit_journal', { handle, unit });
}

File: src-tauri/src/commands/journal.rs
//! A systemd unit's own recent journal lines, for a session already
//! connected.
//!
//! Read-only, matching `ssh::journal`'s own scope.

use tauri::State;

use crate::error::{Error, IpcError};
use crate::ssh::journal;
use crate::ssh::registry::{Registry, SessionHandle};

/// Runs [`journal::command`] for `unit` over `handle`'s connection and
/// parses what came back. A `unit` that does not look like a real systemd
/// unit name yields an empty list without the host ever being asked, the
/// same "nothing to report" shape [`journal::parse`] already answers a
/// permission error or a missing `journalctl` with.
async fn unit_journal(
    registry: &Registry,
    handle: SessionHandle,
    unit: &str,
) -> Result<Vec<String>, Error> {
    let shared = registry.shared(handle).await.ok_or(Error::UnknownHandle)?;

    let Some(command) = journal::command(unit) else {
        return Ok(Vec::new());
    };

    let output = {
        let held = shared.lock().await;
        let connection = held.as_ref().ok_or(Error::UnknownHandle)?;
        connection.run_command(&command).await
    }
    .map_err(Box::new)
    .map_err(Error::Ssh)?;

    Ok(journal::parse(&output.stdout))
}

#[tauri::command]
pub async fn session_unit_journal(
    registry: State<'_, Registry>,
    handle: SessionHandle,
    unit: String,
) -> Result<Vec<String>, IpcError> {
    Ok(unit_journal(&registry, handle, &unit).await?)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[tokio::test]
    async fn an_unknown_handle_is_refused_before_any_command_is_sent() {
        let phantom: SessionHandle = serde_json::from_str("999999").expect("deserializes");
        let registry = Registry::new();

        assert!(matches!(
            unit_journal(&registry, phantom, "ssh.service").await,
            Err(Error::UnknownHandle)
        ));
    }
}
